metadata-ingestion/src/datahub/ingestion/graph/client.py

```diff
@@ -738,32 +751,174 @@ def get_urns_by_filter(
+            variables["scrollId"] = scroll_id
 
-            variables = {
-                "types": types,
-                "query": query,
-                "orFilters": orFilters,
-                "batchSize": batch_size,
-                "scrollId": scroll_id,
-            }
             response = self.execute_graphql(
                 graphql_query,
                 variables=variables,
             )
             data = response["scrollAcrossEntities"]
             scroll_id = data["nextScrollId"]
             for entry in data["searchResults"]:
-                yield entry["entity"]["urn"]
+                yield entry["entity"]
 
             if scroll_id:
                 logger.debug(
                     f"Scrolling to next scrollAcrossEntities page: {scroll_id}"
                 )
 
+    def _add_env_filter(
+        self, env: Optional[str], orFilters: List[Dict[str, List[SearchFilterRule]]]
+    ) -> List[Dict[str, List[SearchFilterRule]]]:
+        if env:
+            # The env filter is a bit more tricky since it's not always stored
+            # in the same place in ElasticSearch.
+            envOrConditions: List[SearchFilterRule] = [
+                # For most entity types, we look at the origin field.
+                {
+                    "field": "origin",
+                    "value": env,
+                    "condition": "EQUAL",
+                },
+                # For containers, we look at the customProperties field.
+                # For any containers created after https://github.com/datahub-project/datahub/pull/8027,
+                # we look for the "env" property. Otherwise, we use the "instance" property.
+                {
+                    "field": "customProperties",
+                    "value": f"env={env}",
+                },
+                {
+                    "field": "customProperties",
+                    "value": f"instance={env}",
+                },
+                # Note that not all entity types have an env (e.g. dashboards / charts).
+                # If the env filter is specified, these will be excluded.
+            ]
+
+            # This matches ALL of the andFilters and at least one of the envOrConditions.
+            orFilters = [
+                {"and": andFilters["and"] + [extraCondition]}
+                for extraCondition in envOrConditions
+                for andFilters in orFilters
+            ]
+
+        return orFilters
+
+    def _add_status_filer(
+        self, status: RemovedStatusFilter, andFilters: List[SearchFilterRule]
+    ) -> None:
+        if status == RemovedStatusFilter.NOT_SOFT_DELETED:
+            # Subtle: in some cases (e.g. when the dataset doesn't have a status aspect), the
+            # removed field is simply not present in the ElasticSearch document. Ideally this
+            # would be a "removed" : "false" filter, but that doesn't work. Instead, we need to
+            # use a negated filter.
+            andFilters.append(
```

Comment:
in general, mutating args is kinda an anti-pattern. is there a way we can avoid doing that?

Reply:
refractored this.